src/grappler/ngraph_optimizer.cc

```diff
@@ -70,20 +70,38 @@ Status NgraphOptimizer::Optimize(tensorflow::grappler::Cluster* cluster,
     return Status::OK();
   }
 
-  // Get the nodes to be skipped
+  // TODO: Find out a better way to preserve feed nodes, init_ops and
+  // keep_ops instead of just skipping those from clustering.
+  // Get nodes to be preserved/skipped
+  std::set<string> nodes_to_preserve;
+
+  // Feed Nodes
+  for (int i = 0; i < item.feed.size(); i++) {
+    nodes_to_preserve.insert(item.feed[i].first);
+  }
+  // Keep Ops
+  for (const string& f : item.keep_ops) {
+    nodes_to_preserve.insert(f);
+  }
```

Comment:
Minor:
`nodes_to_preserve.insert(item.keep_ops.begin(), item.keep_ops.end())`

similarly for `item.init_ops`